syntax = "proto3";

package dfs_service;

import "google/protobuf/timestamp.proto";

service DFSService {


    rpc StoreFile(stream StoreRequest) returns (StoreResponse);
    rpc FetchFile(FetchRequest) returns (stream FetchResponse);
    rpc ListFiles(ListRequest) returns (ListResponse);
    rpc GetAttrs(AttrsRequest) returns (AttrsResponse);
    rpc DeleteFile(DeleteRequest) returns (DeleteResponse);

}

message StoreRequest {
    string filename = 1;
    bytes chunk = 2; // The file data chunk.
}

message StoreResponse {
    bool success = 1; // Whether the file was successfully stored.
    string filename = 2; // The name of the stored file.
    google.protobuf.Timestamp modified_time = 3; // The last modified time of the stored file.
}

message FetchRequest {
    string filename = 1; // The name of the file to fetch.
}

message FetchResponse {
    bytes chunk = 1; // A chunk of the file content.
}

message FileInfo {
    string name = 1; // The name of the file.
    int64 size = 2; // The size of the file in bytes.
    google.protobuf.Timestamp modified_time = 3; // The last modified time of the file.
    google.protobuf.Timestamp creation_time = 4; // The creation time of the file.
}

message ListRequest {}

message ListResponse {
    repeated FileInfo files = 1; // The list of files.
}

message AttrsRequest {
    string filename = 1; // The name of the file.
}


message AttrsResponse {
    FileInfo info = 1; // The attributes of the file.
}


message DeleteRequest {
    string filename = 1; // The name of the file to delete.
}


message DeleteResponse {
    bool success = 1; // Whether the file was successfully deleted.
}

// Add your message types here

